File: C#/567. Permutation in String.cs
/*
 * Link: https://leetcode.com/problems/permutation-in-string/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public bool CheckInclusion(string s1, string s2) {
        var map = new Dictionary<char, int>();
        
        foreach(char c in s1) {
            if (!map.ContainsKey(c)) map.Add(c, 0);
            map[c]++;
        }
        
        int start = 0, end = 0;
        int count = s1.Length;
        
        while(end < s2.Length) {
            if (map.ContainsKey(s2[end]) && map[s2[end]]-- > 0) {
                count--;
            }
            end++;
            
            while(count == 0) {
                if (end - start == s1.Length) return true;
                
                if (map.ContainsKey(s2[start]) && map[s2[start]]++ == 0) {
                    count++;
                }
                start++;
            }
        }
        
        return false;
    }
}

File: Python/567. Permutation in String.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        map, start, end, d = {}, 0, 0, len(s1)
        for c in s1:
                map[c] = 1 if c not in map else map[c] + 1
        while end < len(s2):
            if s2[end] in map:
                if map[s2[end]] > 0:
                    d -= 1
                map[s2[end]] -= 1
            end += 1
            while d == 0:
                if end - start == len(s1):
                    return True
                if s2[start] in map:
                    if map[s2[start]] == 0 :
                        d += 1
                    map[s2[start]] += 1
                start += 1
        return False
        